tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,512 @@
+				 * invalidation log. */
+				ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+				ts_catalog_delete_only(ti->scanrel, tuple);
+				ts_catalog_restore_user(&sec_ctx);
+			}
+
+			if (should_free)
+				heap_freetuple(tuple);
+
+			MemoryContextSwitchTo(oldmctx);
+			MemoryContextReset(state->per_tuple_mctx);
+		}
+	}
+
+	ts_scan_iterator_close(&iterator);
+}
+
+static void
+cagg_invalidations_scan_by_hypertable_init(ScanIterator *iterator, int32 cagg_hyper_id,
+										   LOCKMODE lockmode)
+{
+	*iterator = ts_scan_iterator_create(CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG,
+										lockmode,
+										CurrentMemoryContext);
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(),
+											CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG,
+											CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG_IDX);
+	ts_scan_iterator_scan_key_init(
+		iterator,
+		Anum_continuous_aggs_materialization_invalidation_log_materialization_id,
+		BTEqualStrategyNumber,
+		F_INT4EQ,
+		Int32GetDatum(cagg_hyper_id));
+}
+/*
+ * Clear all cagg invalidations that match a refresh window.
+ *
+ * This function clears all invalidations in the cagg invalidation log that
+ * matches a window. Note that the refresh currently doesn't make use of the
+ * invalidations to optimize the materialization.
+ *
+ * An invalidation entry that gets processed is either completely enclosed
+ * (covered) by the refresh window, or it partially overlaps. In the former
+ * case, the invalidation entry is removed and for the latter case it is
+ * cut. Thus, an entry can either disappear, reduce in size, or be cut in two.
+ *
+ * Note that the refresh window is inclusive at the start and exclusive at the
+ * end.
+ */
+static void
+clear_cagg_invalidations_for_refresh(const CaggInvalidationState *state,
+									 const InternalTimeRange *refresh_window)
+{
+	ScanIterator iterator;
+	int32 cagg_hyper_id = state->cagg.data.mat_hypertable_id;
+
+	/* The scanner defaults to SnapshotSelf, which would show the
+	 * modifications we do (new and updated tuples) within the scan loop
+	 * below. Therefore, we use our own snapshot here. */
+	cagg_invalidations_scan_by_hypertable_init(&iterator, cagg_hyper_id, RowExclusiveLock);
+	iterator.ctx.snapshot = state->snapshot;
+	MemoryContextReset(state->per_tuple_mctx);
+
+	/* Process all invalidations for the continuous aggregate */
+	ts_scanner_foreach(&iterator)
+	{
+		Form_continuous_aggs_materialization_invalidation_log form;
+		MemoryContext oldmctx;
+		InvalidationResult result;
+		bool should_free;
+		HeapTuple tuple;
+
+		oldmctx = MemoryContextSwitchTo(state->per_tuple_mctx);
+		tuple = ts_scan_iterator_fetch_heap_tuple(&iterator, false, &should_free);
+		form = (Form_continuous_aggs_materialization_invalidation_log) GETSTRUCT(tuple);
+		result = cut_invalidation_along_refresh_window(state,
+													   (InvalidationEntry *) form,
```

Comment:
Recasting the `Form` structure to another structure looks dangerous: it is easy to update one and forget to update the other. If they really share representation, then it is probably better to have a single `Form_...` variable for them both and not cast between them.